Type issue with interrupt handler

We have an application where we set the interrupt handler. When running tests, every now and then the following panic pops up:

> interface conversion: interface {} is func(*quickjs.Context, quickjs.Value, []quickjs.Value) quickjs.Value, not quickjs.InterruptHandler

Stack is something like

```
panic: interface conversion: interface {} is func(*quickjs.Context, quickjs.Value, []quickjs.Value) quickjs.Value, not quickjs.InterruptHandler [recovered]
        panic: interface conversion: interface {} is func(*quickjs.Context, quickjs.Value, []quickjs.Value) quickjs.Value, not quickjs.InterruptHandler

goroutine 8 [running, locked to thread]:
testing.tRunner.func1.2({0x1023e2440, 0x1400067f4a0})
        /opt/homebrew/Cellar/go/1.24.2/libexec/src/testing/testing.go:1734 +0x1ac
testing.tRunner.func1()
        /opt/homebrew/Cellar/go/1.24.2/libexec/src/testing/testing.go:1737 +0x334
panic({0x1023e2440?, 0x1400067f4a0?})
        /opt/homebrew/Cellar/go/1.24.2/libexec/src/runtime/panic.go:792 +0x124
github.com/buke/quickjs-go.goInterruptHandler(0x140007f8b28?, 0x1009f45dc?)
        /[me]/go/pkg/mod/github.com/buke/quickjs-go@v0.4.15/bridge.go:71 +0x58
github.com/buke/quickjs-go._Cfunc_JS_Call(0x154765b40, {{0x50, 0x31, 0x78, 0x54, 0x1, 0x0, 0x0, 0x0}, 0xffffffffffffffff}, ...)
        _cgo_gotypes.go:291 +0x34
github.com/buke/quickjs-go.Value.Call.func2(...)
        /[me]/go/pkg/mod/github.com/buke/quickjs-go@v0.4.15/value.go:254
github.com/buke/quickjs-go.Value.Call({0x140008dd710, {{0xe0, 0x6a, 0x76, 0x54, 0x1, 0x0, 0x0, 0x0}, 0xffffffffffffffff}}, ...)
        /[me]/go/pkg/mod/github.com/buke/quickjs-go@v0.4.15/value.go:254 +0x388
[...]
```

I don't know where this `func(*quickjs.Context, quickjs.Value, []quickjs.Value) quickjs.Value` is coming from. The interrupt handler has certainly the correct signature.

However, looking into `func (ctx *Context) SetInterruptHandler(handler InterruptHandler)` I wonder why `unsafe.Pointer(&handlerArgs)` is safe. Isn't https://github.com/buke/quickjs-go/blob/v0.4.15/context.go#L217-L220 turning the shortly living `handlerArgs := C.handlerArgs`  into a long living pointer? Can't Go just move `handlerArgs` or delete it?

--- a/context.go
+++ b/context.go
@@ -3,6 +3,7 @@ package quickjs
 import (
 	"fmt"
 	"os"
+	"runtime"
 	"runtime/cgo"
 	"unsafe"
 )
@@ -20,6 +21,7 @@ type Context struct {
 	globals    *Value
 	proxy      *Value
 	asyncProxy *Value
+	pinner     runtime.Pinner
 }
 
 // Runtime returns the runtime of the context.
@@ -42,6 +44,8 @@ func (ctx *Context) Close() {
 	}
 
 	C.JS_FreeContext(ctx.ref)
+
+	ctx.pinner.Unpin()
 }
 
 // Null return a null value.
@@ -214,10 +218,14 @@ type InterruptHandler func() int
 
 // SetInterruptHandler sets a interrupt handler.
 func (ctx *Context) SetInterruptHandler(handler InterruptHandler) {
-	handlerArgs := C.handlerArgs{
+	handlerArgsPtr := &C.handlerArgs{
 		fn: (C.uintptr_t)(cgo.NewHandle(handler)),
 	}
-	C.SetInterruptHandler(ctx.runtime.ref, unsafe.Pointer(&handlerArgs))
+
+	// Ensure the C.handlerArgs instance is never moved to a different place or GCed.
+	ctx.pinner.Pin(handlerArgsPtr)
+
+	C.SetInterruptHandler(ctx.runtime.ref, unsafe.Pointer(handlerArgsPtr))
 }
 
 // Atom returns a new Atom value with given string.

--- a/context.go
+++ b/context.go
@@ -3,7 +3,6 @@ package quickjs
 import (
 	"fmt"
 	"os"
-	"runtime"
 	"runtime/cgo"
 	"unsafe"
 )
@@ -21,7 +20,6 @@ type Context struct {
 	globals    *Value
 	proxy      *Value
 	asyncProxy *Value
-	pinner     runtime.Pinner
 }
 
 // Runtime returns the runtime of the context.
@@ -44,8 +42,6 @@ func (ctx *Context) Close() {
 	}
 
 	C.JS_FreeContext(ctx.ref)
-
-	ctx.pinner.Unpin()
 }
 
 // Null return a null value.
@@ -217,15 +213,10 @@ func (ctx *Context) AsyncFunction(asyncFn func(ctx *Context, this Value, promise
 type InterruptHandler func() int
 
 // SetInterruptHandler sets a interrupt handler.
+//
+// Deprecated: Use SetInterruptHandler on runtime instead
 func (ctx *Context) SetInterruptHandler(handler InterruptHandler) {
-	handlerArgsPtr := &C.handlerArgs{
-		fn: (C.uintptr_t)(cgo.NewHandle(handler)),
-	}
-
-	// Ensure the C.handlerArgs instance is never moved to a different place or GCed.
-	ctx.pinner.Pin(handlerArgsPtr)
-
-	C.SetInterruptHandler(ctx.runtime.ref, unsafe.Pointer(handlerArgsPtr))
+	ctx.runtime.SetInterruptHandler(handler)
 }
 
 // Atom returns a new Atom value with given string.

--- a/quickjs_test.go
+++ b/quickjs_test.go
@@ -696,7 +696,7 @@ func TestAsyncFunction(t *testing.T) {
 	require.EqualValues(t, "Hello Async", ret3.String())
 }
 
-func TestSetInterruptHandler(t *testing.T) {
+func TestSetInterruptHandlerContext(t *testing.T) {
 	rt := quickjs.NewRuntime()
 	defer rt.Close()
 
@@ -712,6 +712,57 @@ func TestSetInterruptHandler(t *testing.T) {
 		return 0
 	})
 
+	ret, err := ctx.Eval(`while(true){}`)
+	defer ret.Free()
+
+	assert.Error(t, err, "expected interrupted by quickjs")
+	require.Equal(t, "InternalError: interrupted", err.Error())
+}
+
+// This test shows how the interrupt handler is not specific to the context.
+// So it is cleaner to set it on the runtime instead.
+func TestSetInterruptHandlerContextIsSetToRuntime(t *testing.T) {
+	rt := quickjs.NewRuntime()
+	defer rt.Close()
+
+	ctx1 := rt.NewContext()
+	defer ctx1.Close()
+	ctx2 := rt.NewContext()
+	defer ctx2.Close()
+
+	startTime := time.Now().Unix()
+
+	// Set it to context 1
+	ctx1.SetInterruptHandler(func() int {
+		if time.Now().Unix()-startTime > 1 {
+			return 1
+		}
+		return 0
+	})
+
+	// use it in context 2
+	ret, err := ctx2.Eval(`while(true){}`)
+	defer ret.Free()
+
+	assert.Error(t, err, "expected interrupted by quickjs")
+	require.Equal(t, "InternalError: interrupted", err.Error())
+}
+
+func TestSetInterruptHandlerRuntime(t *testing.T) {
+	rt := quickjs.NewRuntime()
+	defer rt.Close()
+
+	startTime := time.Now().Unix()
+
+	rt.SetInterruptHandler(func() int {
+		if time.Now().Unix()-startTime > 1 {
+			return 1
+		}
+		return 0
+	})
+
+	ctx := rt.NewContext()
+	defer ctx.Close()
 	ret, err := ctx.Eval(`while(true){}`)
 	defer ret.Free()
 

--- a/runtime.go
+++ b/runtime.go
@@ -7,6 +7,7 @@ package quickjs
 import "C"
 import (
 	"runtime"
+	"runtime/cgo"
 	"unsafe"
 )
 
@@ -14,6 +15,7 @@ import (
 type Runtime struct {
 	ref     *C.JSRuntime
 	options *Options
+	pinner  runtime.Pinner
 }
 
 type Options struct {
@@ -112,6 +114,7 @@ func (r Runtime) RunGC() {
 // Close will free the runtime pointer.
 func (r Runtime) Close() {
 	C.JS_FreeRuntime(r.ref)
+	r.pinner.Unpin()
 }
 
 // SetCanBlock will set the runtime's can block; default is true
@@ -141,6 +144,18 @@ func (r Runtime) SetMaxStackSize(stack_size uint64) {
 // SetExecuteTimeout will set the runtime's execute timeout; default is 0
 func (r Runtime) SetExecuteTimeout(timeout uint64) {
 	C.SetExecuteTimeout(r.ref, C.time_t(timeout))
+}
+
+// SetInterruptHandler sets a interrupt handler.
+func (r *Runtime) SetInterruptHandler(handler InterruptHandler) {
+	handlerArgsPtr := &C.handlerArgs{
+		fn: (C.uintptr_t)(cgo.NewHandle(handler)),
+	}
+
+	// Ensure the C.handlerArgs instance is never moved to a different place or GCed.
+	r.pinner.Pin(handlerArgsPtr)
+
+	C.SetInterruptHandler(r.ref, unsafe.Pointer(handlerArgsPtr))
 }
 
 // NewContext creates a new JavaScript context.

